lib/srv/db/cassandra/protocol/conn.go

```diff
@@ -0,0 +1,213 @@
+type Conn struct {
+	net.Conn
+	segmentCode       segment.Codec
+	frameCodec        frame.RawCodec
+	modernLayoutRead  bool
+	modernLayoutWrite bool
+}
+
+// ReadPacket is used to read packet from the connection.
+func (c *Conn) ReadPacket() (*Packet, error) {
+	var buff bytes.Buffer
+	tr := io.TeeReader(c.Conn, &buff)
+
+	fr, err := c.readWithModernLayout(tr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	c.maybeSwitchToModernLayout(fr)
+
+	return &Packet{
+		raw:   buff,
+		frame: fr,
+	}, nil
+}
+
+// WriteFrame is used to write frame to the connection.
+func (c *Conn) WriteFrame(outgoing *frame.Frame) error {
+	if err := c.writeFrameWithModernLayout(outgoing); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if startup, ok := outgoing.Body.Message.(*message.Startup); ok {
+		compression := startup.GetCompression()
+		c.frameCodec = frame.NewRawCodecWithCompression(client.NewBodyCompressor(compression))
+		c.segmentCode = segment.NewCodecWithCompression(client.NewPayloadCompressor(compression))
+	}
+	c.maybeSwitchToModernLayout(outgoing)
+	return nil
+}
+
+// readWithModernLayout is used to read frame from the connection.
+// If the connection is using modern framing layout, it will read segments.
+// Otherwise, it will read frames.
+func (c *Conn) readWithModernLayout(r io.Reader) (*frame.Frame, error) {
+	if c.modernLayoutRead {
+		v, err := c.readSegment(r)
+		return v, trace.Wrap(err)
+	}
+	fr, err := c.readFrame(r)
+	return fr, trace.Wrap(err)
+}
+
+// readFrame is used to read frame from the connection.
+// If read frame is a Startup frame, it will switch to modern framing layout and
+// update codec to use modern framing layout.
+func (c *Conn) readFrame(r io.Reader) (*frame.Frame, error) {
+	fr, err := c.frameCodec.DecodeFrame(r)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if startup, ok := fr.Body.Message.(*message.Startup); ok {
+		compression := startup.GetCompression()
+		c.frameCodec = frame.NewRawCodecWithCompression(client.NewBodyCompressor(compression))
+		c.segmentCode = segment.NewCodecWithCompression(client.NewPayloadCompressor(compression))
+		// If moderate framing layout is supported all received from a client after Startup message should
+		// use segment encoding.
+		c.modernLayoutRead = fr.Header.Version.SupportsModernFramingLayout()
+	}
+
+	return fr, nil
+}
+
+// readSegment is used to read segments from the connection.
+// If frame is not self-contained, segments are split into multiple frames.
+// Read the segments till received bodyLength bytes.
+func (c *Conn) readSegment(r io.Reader) (*frame.Frame, error) {
+	previousSegment := bytes.Buffer{}
+	expectedSegmentSize := 0
+
+	for {
+		seg, err := c.segmentCode.DecodeSegment(r)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if seg.Header.IsSelfContained {
+			fr, err := c.readFrame(bytes.NewReader(seg.Payload.UncompressedData))
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			return fr, nil
+		}
```

Comment:
When self contained, the segment may have one or *more* frames/envelopes?

Reply:
Quote

>  If self contained, then the payload includes one or **more** complete envelopes

Reply:
This function was borrow from: https://github.com/datastax/go-cassandra-native-protocol/blob/5e8aad05cf90f30499e5873f70b37dbc175306fd/client/client.go#L323
I assume that the "more complete envelopes" case  is handled by library internals.

Reply:
@smallinsky Are you sure that @greedy52 is not right? This is the implementation that I see https://github.com/datastax/go-cassandra-native-protocol/blob/5e8aad05cf90f30499e5873f70b37dbc175306fd/client/server.go#L549-L557
it reads multiple frames from one envelope until the buffer is empty. Yours from that I see reads only one.

@greedy52 Sorry for being not specific. I just wanted to say that "self-contained" means that all data that we need to encode the message should be already in the buffer, whereas not self-contained requires reading more data from the network.

Reply:
Thanks for pointing this. Yes your are right the `self-contained` should have spearer implementation where multiple segment can contains severals frames. I have added additional flow `readSelfContainedSegment`